src/Utilities/TrackedDependencies/CanonicalTrackedOutputFiles.cs

```diff
@@ -767,8 +764,18 @@ private void RemoveDependenciesFromEntryIfMissing(string rootingMarker)
                 {
                     if (keyIndex++ > 0)
                     {
-                        // If we are ignoring missing files, then only record those that exist
-                        if (FileUtilities.FileExistsNoThrow(file))
+                        // Record whether or not each file exists and cache it.
+                        // We do this to save time (On^2), at the expense of data O(n).
+                        bool inFileCache = fileCache.TryGetValue(file, out bool fileExists);
+
+                        // Have we cached the file yet? If not, cache whether or not it exists.
+                        if (!inFileCache)
+                        {
+                            fileCache.Add(file, FileUtilities.FileExistsNoThrow(file));
```

Comment:
Same comment as in the other file. The result of `FileExistsNoThrow()` is cached but it's not actually used below.